File: Data Structures & Algorithms/largest-rectangle-in-histogram/submission-1.py
class Solution:
    def largestRectangleArea(self, heights: List[int]) -> int:
        n = len(heights)
        st = []
        res = 0

        for i in range(n):
            h = heights[i]
            start = i
            while st and h < st[-1][1]:
                start = st[-1][0]
                area = st[-1][1] * (i - st[-1][0])
                res = max(res, area)
                st.pop()
            
            st.append((start, h))
        
        while st:
            area = st[-1][1] * (n - st[-1][0])
            res = max(res, area)
            st.pop()
        
        return res

File: Data Structures & Algorithms/longest-substring-without-duplicates/submission-1.cpp
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        int n = s.size();
        unordered_map<char, int> m;
        int res = 0;

        int lo = 0, hi = 0;
        while(hi<n){
            while(m[s[hi]]>0){
                // cout << "   s[hi]:" << s[hi] << " m[s[hi]]:" << m[s[hi]] << endl;
                m[s[lo]]--;
                lo++;
            }
            // cout << "lo: " << lo << " hi:" << hi << endl;

            m[s[hi]]++;
            res = max(res, hi-lo+1);
            hi++;
        }

        return res;
    }
};
